File: sop-java-picocli/src/main/java/sop/cli/picocli/commands/InlineDetachCmd.java
// SPDX-FileCopyrightText: 2021 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package sop.cli.picocli.commands;

import picocli.CommandLine;
import sop.Signatures;
import sop.cli.picocli.SopCLI;
import sop.exception.SOPGPException;
import sop.operation.InlineDetach;

import java.io.IOException;
import java.io.OutputStream;

@CommandLine.Command(name = "inline-detach",
        resourceBundle = "msg_inline-detach",
        exitCodeOnInvalidInput = SOPGPException.UnsupportedOption.EXIT_CODE)
public class InlineDetachCmd extends AbstractSopCmd {

    @CommandLine.Option(
            names = {"--signatures-out"},
            paramLabel = "SIGNATURES")
    String signaturesOut;

    @CommandLine.Option(names = "--no-armor",
            negatable = true)
    boolean armor = true;

    @Override
    public void run() {
        InlineDetach inlineDetach = throwIfUnsupportedSubcommand(
                SopCLI.getSop().inlineDetach(), "inline-detach");

        throwIfOutputExists(signaturesOut);
        throwIfMissingArg(signaturesOut, "--signatures-out");

        if (!armor) {
            inlineDetach.noArmor();
        }

        try (OutputStream outputStream = getOutput(signaturesOut)) {
            Signatures signatures = inlineDetach
                    .message(System.in).writeTo(System.out);
            signatures.writeTo(outputStream);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }
}

File: sop-java-picocli/src/main/kotlin/sop/cli/picocli/commands/InlineDetachCmd.kt
// SPDX-FileCopyrightText: 2023 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package sop.cli.picocli.commands

import java.io.IOException
import java.lang.RuntimeException
import picocli.CommandLine.Command
import picocli.CommandLine.Option
import sop.cli.picocli.SopCLI
import sop.exception.SOPGPException

@Command(
    name = "inline-detach",
    resourceBundle = "msg_inline-detach",
    exitCodeOnInvalidInput = SOPGPException.UnsupportedOption.EXIT_CODE)
class InlineDetachCmd : AbstractSopCmd() {

    @Option(names = [OPT_SIGNATURES_OUT], paramLabel = "SIGNATURES")
    var signaturesOut: String? = null

    @Option(names = [OPT_NO_ARMOR], negatable = true) var armor: Boolean = true

    override fun run() {
        val inlineDetach =
            throwIfUnsupportedSubcommand(SopCLI.getSop().inlineDetach(), "inline-detach")

        throwIfOutputExists(signaturesOut)
        throwIfMissingArg(signaturesOut, OPT_SIGNATURES_OUT)

        if (!armor) {
            throwIfUnsupportedOption(OPT_NO_ARMOR) { inlineDetach.noArmor() }
        }

        try {
            getOutput(signaturesOut).use { sigOut ->
                inlineDetach
                    .message(System.`in`)
                    .writeTo(System.out) // message out
                    .writeTo(sigOut) // signatures out
            }
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
    }

    companion object {
        const val OPT_NO_ARMOR = "--no-armor"
        const val OPT_SIGNATURES_OUT = "--signatures-out"
    }
}
